File: Android/app/src/main/java/leapfrog_inc/icchi/Http/Requester/UserRequester.java
package leapfrog_inc.icchi.Http.Requester;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;

import leapfrog_inc.icchi.Function.Constants;
import leapfrog_inc.icchi.Function.SaveData;
import leapfrog_inc.icchi.Http.HttpManager;

/**
 * Created by Leapfrog-Software on 2018/01/27.
 */

public class UserRequester {

    public enum AgeType {
        u20,
        s20,
        s30,
        s40,
        s50,
        o60;

        public static AgeType init(String string) {
            if (string.equals("0")) {
                return AgeType.u20;
            } else if (string.equals("1")) {
                return AgeType.s20;
            } else if (string.equals("2")) {
                return AgeType.s30;
            } else if (string.equals("3")) {
                return AgeType.s40;
            } else if (string.equals("4")) {
                return AgeType.s50;
            } else if (string.equals("5")) {
                return AgeType.o60;
            } else {
                return null;
            }
        }

        public String convert() {
            if (this == AgeType.u20) {
                return "0";
            } else if (this == AgeType.s20) {
                return "1";
            } else if (this == AgeType.s30) {
                return "2";
            } else if (this == AgeType.s40) {
                return "3";
            } else if (this == AgeType.s50) {
                return "4";
            } else if (this == AgeType.o60) {
                return "5";
            } else {
                return "";
            }
        }

        public String display() {
            if (this == AgeType.u20) {
                return "20歳未満";
            } else if (this == AgeType.s20) {
                return "20代";
            } else if (this == AgeType.s30) {
                return "30代";
            } else if (this == AgeType.s40) {
                return "40代";
            } else if (this == AgeType.s50) {
                return "50代";
            } else if (this == AgeType.s50) {
                return "60歳以上";
            } else {
                return "";
            }
        }
    }

    public enum GenderType {
        male,
        female;

        public static GenderType init(String string) {
            if (string.equals("0")) {
                return GenderType.male;
            } else if (string.equals("1")) {
                return GenderType.female;
            } else {
                return null;
            }
        }

        public String convert() {
            if (this == GenderType.male) {
                return "0";
            } else if (this == GenderType.female) {
                return "1";
            } else {
                return "";
            }
        }

        public String display() {
            if (this == GenderType.male) {
                return "男性";
            } else if (this == GenderType.male) {
                return "女性";
            } else {
                return "";
            }
        }
    }

    public static class UserData {

        public String userId = "";
        public String name = "";
        public AgeType age = null;
        public GenderType gender = null;
        public ArrayList<String> likes = new ArrayList<String>();
        public ArrayList<String> hates = new ArrayList<String>();
        public String image = "";
        public String fbLink = "";

        static public UserData create(JSONObject json) {

            try {
                String userId = json.getString("userId");
                String name = json.getString("name");
                String age = json.getString("age");
                String gender = json.getString("gender");
                String[] likes = json.getString("likes").split("-");
                String[] hates = json.getString("hates").split("-");
                String image = json.getString("image");
                String fbLink = json.getString("fbLink");

                UserData userData = new UserData();
                userData.userId = userId;
                userData.name = name;
                userData.age = AgeType.init(age);
                userData.gender = GenderType.init(gender);

                userData.likes = new ArrayList<String>();
                for (int i = 0; i < likes.length; i++) {
                    userData.likes.add(likes[i]);
                }

                userData.hates = new ArrayList<String>();
                for (int i = 0; i < hates.length; i++) {
                    userData.hates.add(hates[i]);
                }

                userData.image = image;
                userData.fbLink = fbLink;

                return userData;

            } catch(Exception e) {}

            return null;
        }
    }

    private static UserRequester requester = new UserRequester();

    private UserRequester(){}

    public static UserRequester getInstance(){
        return requester;
    }

    private ArrayList<UserData> mDataList = new ArrayList<UserData>();

    public void fetch(final UserRequesterCallback callback){

        HttpManager httpManager = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        if (ret.equals("0")) {
                            JSONArray jsonArray = jsonObject.getJSONArray("users");
                            ArrayList<UserData> dataList = new ArrayList<UserData>();
                            for (int i = 0; i < jsonArray.length(); i++) {
                                UserData userData = UserData.create(jsonArray.getJSONObject(i));
                                if (userData != null) {
                                    dataList.add(userData);
                                }
                            }
                            mDataList = dataList;
                            callback.didReceiveData(true);
                            return;
                        }
                    } catch(Exception e) {}
                }
                callback.didReceiveData(false);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=getUser");
        httpManager.execute(Constants.ServerRootUrl, "POST", param.toString());
    }

    public interface UserRequesterCallback {
        void didReceiveData(boolean result);
    }

    public ArrayList<UserData> getDataList() {
        return mDataList;
    }

    public UserData query(String userId) {

        for (int i = 0; i < mDataList.size(); i++) {
            if (mDataList.get(i).userId.equals(userId)) {
                return mDataList.get(i);
            }
        }
        return null;
    }

    public int queryMatch(String targetId) {

        UserData targetUserData = query(targetId);
        UserData myUserData = query(SaveData.getInstance().userId);
        if ((targetUserData == null) || (myUserData == null)) {
            return 0;
        }

        int match = 0;

        for (int i = 0; i < targetUserData.likes.size(); i++) {
            String itemId = targetUserData.likes.get(i);
            if (myUserData.likes.contains(itemId)) {
                if (isMinorItem(itemId)) {
                    match += 20;
                } else {
                    match += 5;
                }
            }
        }
        for (int i = 0; i < targetUserData.hates.size(); i++) {
            String itemId = targetUserData.hates.get(i);
            if (myUserData.likes.contains(itemId)) {
                if (isMinorItem(itemId)) {
                    match += 20;
                } else {
                    match += 5;
                }
            }
        }
        if (match > 100) {
            match = 100;
        }
        return match;
    }

    private boolean isMinorItem(String itemId) {

        int count = 0;
        for (int i = 0; i < mDataList.size(); i++) {
            UserData userData = mDataList.get(i);
            if (userData.likes.contains(itemId)) {
                count++;
            } else if (userData.hates.contains(itemId)) {
                count++;
            }
        }
        if (count * 100 < mDataList.size()) {
            return true;
        } else {
            return false;
        }
    }
}

File: iOS/Icchi/Icchi/Http/Requester/UserRequester.swift
//
//  UserRequester.swift
//  Icchi
//
//  Created by oonaka on 2018/02/15.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

class UserRequester {
    
    // 年齢
    enum AgeType: String, Codable {
        case u20 = "0"
        case s20 = "1"
        case s30 = "2"
        case s40 = "3"
        case s50 = "4"
        case o60 = "5"
        
        func convert() -> String {
            if (self == AgeType.u20
                || self == AgeType.s20
                || self == AgeType.s30
                || self == AgeType.s40
                || self == AgeType.s50
                || self == AgeType.o60
                ) {
                return self.rawValue;
            } else {
                return "";
            }
        }
        
        func display() -> String {
            if (self == AgeType.u20) {
                return "20歳未満";
            } else if (self == AgeType.s20) {
                return "20代";
            } else if (self == AgeType.s30) {
                return "30代";
            } else if (self == AgeType.s40) {
                return "40代";
            } else if (self == AgeType.s50) {
                return "50代";
            } else if (self == AgeType.o60) {
                return "60歳以上";
            } else {
                return "";
            }
        }
    }
    
    // 性別
    enum GenderType: String, Codable {
        case male = "0"
        case female = "1"
    
        func convert() -> String {
            if (self == GenderType.male
                || self == GenderType.female) {
                return self.rawValue;
            } else {
                return "";
            }
        }
        
        func display() -> String {
            if (self == GenderType.male) {
                return "男性";
            } else if (self == GenderType.female) {
                return "女性";
            } else {
                return "";
            }
        }
    }
    
    // レスポンスデータ型
    private struct fetchResult : Codable {
        var result:Bool
        var users:[UserData]?
        private enum CodingKeys: String, CodingKey {
            case result
            case users
        }
        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            result = NSString(string:try values.decode(String.self, forKey: .result)).isEqual(to: "0")
            users = try values.decode([UserData].self, forKey: .users)
        }
    }
    
    // ユーザーデータ
    struct UserData : Codable {
        var userId:String? = nil
        var name:String? = nil
        var age:AgeType? = nil
        var gender:GenderType? = nil
        var likes:[String]? = [""]
        var hates:[String]? = [""]
        var image:String? = ""
        var fbLink:String? = ""

        mutating public func likesAddChange(like:String) {
            self.likes?.append(like)
        }
        mutating public func hatesAddChange(hate:String) {
            self.hates?.append(hate)
        }
        
        private enum CodingKeys: String, CodingKey {
            case userId
            case name
            case age
            case gender
            case likes
            case hates
            case image
            case fbLink
        }

        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            self.userId = try? values.decode(String.self, forKey: .userId)
            
            let rawName = try? values.decode(String.self, forKey: .name)
            self.name = rawName?.base64Decode()
            
            self.age = AgeType(rawValue:try values.decode(String.self, forKey: .age))
            self.gender = GenderType(rawValue:try values.decode(String.self, forKey: .gender))
            self.likes = try? values.decode(String.self, forKey: .likes).components(separatedBy: "-")
            self.hates = try? values.decode(String.self, forKey: .hates).components(separatedBy: "-")
            self.image = try? values.decode(String.self, forKey: .image)
            self.fbLink = try? values.decode(String.self, forKey: .fbLink)
        }
    }
    
    // シングルトン
    static let sharedManager = UserRequester()
    
    private init() {}
    
    // ユーザーデータ配列
    fileprivate(set) var mDataList:[UserData] = []
    
    /** データ取得 */
    func fetch(completion: @escaping ((Bool) -> ())) {
        
        // リクエストデータ作成
        let params = ["command": "getUser"]
        // リクエスト実施
        HttpManager.post(url: Constants.ServerApiUrl, params: params) { (result:Bool, data:Data?) in
            
            // レスポンスチェック
            guard result, let data = data else {
                // 失敗
                completion(false)
                return
            }
            
            do {
                // レスポンス取得
                let result = try JSONDecoder().decode(fetchResult.self, from: data)
                
                // ユーザーIDが存在するリストを登録
                if let users = result.users {
                    self.mDataList = users.filter({ (userData) -> Bool in
                        if let userId = userData.userId {
                            return userId.count > 0
                        }
                        return false
                    })
                }
                
                // 成功
                completion(result.result)
            }
            catch {
                // 失敗
                completion(false)
            }
        }
    }
    
    /** データ取得 */
    func query(_ userId:String) -> UserData? {
        return mDataList.first(where: { return $0.userId == userId})
    }
    
    /** 一致度を取得 */
    func queryMatch(_ targetId:String?) -> Int {
        
        guard let targetId = targetId else {
            return 0;
        }
        
        guard let targetUserData = self.query(targetId)
            , let myUserData = query(SaveData.shared.userId) else {
            return 0;
        }
        
        let pointPerItem = ParameterRequester.sharedManager.mPointPerItem;
        let pointPerMinorItem = ParameterRequester.sharedManager.mPointPerMinorItem;
        
        var match = 0;
        
        targetUserData.likes?.forEach({ itemId in
            if let likes = myUserData.likes, likes.contains(itemId) {
                if self.isMinorItem(itemId) {
                    match += pointPerMinorItem
                }
                else {
                    match += pointPerItem
                }
            }
        })
        
        targetUserData.hates?.forEach({ itemId in
            if let hates = myUserData.hates, hates.contains(itemId) {
                if self.isMinorItem(itemId) {
                    match += pointPerMinorItem
                }
                else {
                    match += pointPerItem
                }
            }
        })
        
        if (match > 100) {
            match = 100;
        }
        return match;
    }
    
    func isMinorItem(_ itemId:String?) -> Bool {
    
        guard let itemId = itemId else {
            return false
        }
        
        var count:Int = 0;
        
        mDataList.forEach { userData in
            
            if let likes = userData.likes, likes.contains(itemId) {
                count+=1;
            }
            if let hates = userData.hates, hates.contains(itemId) {
                count+=1;
            }
        }
        
        if (count * 100 * ParameterRequester.sharedManager.mMinorThreshold < mDataList.count) {
            return true;
        } else {
            return false;
        }
    }
}

